### PR TITLE
BSM-118: Added documentation for BSM feature

### PR DESCRIPTION
**Pull request by fooker**
JIRA: http://issues.opennms.org/browse/BSM-118
Bamboo: http://bamboo.internal.opennms.com:8085/browse/OPENNMS-ONMS439/latest

### CODE REVIEW

File: opennms-doc/guide-admin/src/asciidoc/index.adoc [hunk 1]
```diff
@@ -133,13 +133,9 @@ include::text/provisioning/single-node.adoc[]
 include::text/provisioning/fine-grained.adoc[]
 include::text/provisioning/api-examples.adoc[]
 
-[ga-business-service-monitoring]
+[[ga-bsm-admin]]
 == Business Service Monitoring
-include::text/bsm/introduction.adoc[]
```
Comment (j-white) [inline opennms-doc/guide-admin/src/asciidoc/index.adoc]: introduction.adoc and severity.adoc were removed from the index, but are still present in the source tree

**Comment by mvrueden** [comment]
In PR https://github.com/OpenNMS/opennms/pull/563 the guide-admin changed. This must be considered when merging either this PR or PR https://github.com/OpenNMS/opennms/pull/563 to `features/bsm`.

File: opennms-doc/guide-admin/src/asciidoc/index.adoc (hunk 1 above)
Comment (christianpape) [inline opennms-doc/guide-admin/src/asciidoc/index.adoc]: You are right. We forgot to remove these two files. They are deleted now.

File: opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+The status of a managed entity is determined based on related events, alarms and outages.
+However, this approach treats every element of the system individually and there is no concept of relationships between different nodes or services in the service assurance layer.
+
+The _Business Service Monitoring_ (BSM) feature adds the ability to model relationships between managed entities.
+It combines sets of managed entities into hierarchically structured _Business Services_ that represent a high level view on its elements.
+The _Business Service'_ operational status is calculated from the alarm states of its associated elements determined by the service assurance layer.
+Of course, the operational status of a _Business Service_ can be checked at any time.
+If the operational status of a _Business Service_ is changing corresponding events are fired.
+
+==== Structure of a Business Service
```
Comment (indigo423) [inline opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc]: please add new lines after heading to be consistent

File: opennms-doc/guide-development/src/asciidoc/text/bsm/bsm-development.adoc [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+==== Business Service Monitoring
+Every aspect of the _Business Service Monitoring_ feature can be controlled via a ReST API.
+The API's endpoint for managing _Business Services_ is located at `/opennms/api/v2/business-services`.
+It supports _XML_ content to represent the _Business Services_. The responses generated by the ReST API
+do also include `location` elements that contain references to other entities managed by the API.
+The _Business Service_ response data model for the ReST API has the following basic structure:
+
+.Sample _Business Service_ details response XML representation
+[source,XML]
```
Comment (indigo423) [inline opennms-doc/guide-development/src/asciidoc/text/bsm/bsm-development.adoc]: please fix indentation starts at column 0.

File: opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+Business Services can contain attributes, _Reduction Keys_, references to monitored services and/or other business services.
+
+===== Reduction Keys as part of a Business Service
```
Comment (indigo423) [inline opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc]: please add new lines after heading to be consistent

File: opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+_Business Services_ enables users to correlate the alarm based status of managed entities into its operational status.
+_Business Services_ can contain a list of _Reduction Keys_ that are used to identify the relevant alarms to summarize.
+_Reduction Keys_ can be added to business services as plain text.
+
+====== Monitored services as part of a Business Service
```
Comment (indigo423) [inline opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc]: please add new lines after heading to be consistent

File: opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+_Business Services_ can contain references to monitored services (node/interface/service combination).
+A referenced monitored service is called _IP Service_ in the _Business Service_ architecture.
+_IP Services_ in a business service are interpreted as two _Reduction Keys_:
+
+* `uei.opennms.org/nodes/nodeLostService::%nodeID%:%ipAddress%:%serviceName%`
+-- This _Reduction Key_ matches the alarms fired if the monitored service goes down.
+* `uei.opennms.org/nodes/nodeDown::%nodeID%`
+-- This _Reduction Key_ matches the alarms fired if the entire node providing the monitored-service goes down.
+
+====== Business Services as part of Business Services
```
Comment (indigo423) [inline opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc]: please add new lines after heading to be consistent

File: opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+_Business services_ can form relationships woth other _Business Services_.
+This offers the ability to create hierarchically structures of _Business Services_.
+To determine the operational status of a enclosing _Business Service_ the operational status of all its enclosed _Business Services_ is summarized.
+
+====== Attributes as part of Business Services
```
Comment (indigo423) [inline opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc]: please add new lines after heading to be consistent

Section with just one sentence?

File: opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+Additionally, each _Business Services_ can contain a set of key value attributes.
+
+==== Operational status of a Business Service
```
Comment (indigo423) [inline opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc]: please add new lines after heading to be consistent

File: opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+For this operation the following _Reduce Functions_ are available:
+
+.Status calculation _Reduce Functions_
+[options="header, autowidth"]
+|===
+| Name            | Description
+| `Most Critical` | Uses the value of the most critical status.
+| `Threshold`     | Uses the highest severity found more often than the given threshold.
+|===
+
```
Comment (indigo423) [inline opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc]: please remove double new line

File: opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+
+===== Operational status changes
```
Comment (indigo423) [inline opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc]: please add new lines after heading to be consistent

File: opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+If a _Business Service_ changes its operational status an _OpenNMS_ event of the type `uei.opennms.org/bsmd/serviceOperationalStatusChanged` is fired.
+The log message of the event contains the name of the _Business Service_.
+Additionally, the previous operational status and the new operational status are included.
+Please note, that this event has no node, interface or service associated.
+
+===== The Business Service Daemon
```
Comment (indigo423) [inline opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc]: please add new lines after heading to be consistent

File: opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+Internally the _Business Service Monitoring_ feature is driven by the _Business Service Daemon_ (bsmd).
+It is responsible for tracking the operational status of all _Business Services_ and for sending events in case of operational status changes.
+Every time the configuration of a _Business Service_ is changed a reload of the daemon's configuration is required.
+This includes changes like the name of the _Business Service_ or its attributes as well as changes regarding the _Reduction Keys_, contained _Business Services_ or _IP Services_.
+
+====== Reloading Business Service Daemon configuration
```
Comment (indigo423) [inline opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc]: please add new lines after heading to be consistent

File: opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+A configuration reload can be initiated by an `uei.opennms.org/internal/reloadDaemonConfig` event with the parameter `daemonName` set to `bsmd`.
+If the reload of the configuration is done an event of type `uei.opennms.org/internal/reloadDaemonConfigSuccessful` is fired.
+The following options to trigger a configuration reload are available:
+
+- Use the `Reload` button on the `Manage Business Services` page in the administration section of the web UI.
+- Send the `reloadDaemonConfig` event directly to trigger the reload
+* Use the `send-event.pl` script
```
Comment (indigo423) [inline opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc]: Add line break between text and source code

File: opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc (hunk 7 above)
Comment (mvrueden) [inline opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc]: Yes. It is a place holder for more to come.

File: opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+It is responsible for tracking the operational status of all _Business Services_ and for sending events in case of operational status changes.
+Every time the configuration of a _Business Service_ is changed a reload of the daemon's configuration is required.
+This includes changes like the name of the _Business Service_ or its attributes as well as changes regarding the _Reduction Keys_, contained _Business Services_ or _IP Services_.
+
+====== Reloading Business Service Daemon configuration
+A configuration reload can be initiated by an `uei.opennms.org/internal/reloadDaemonConfig` event with the parameter `daemonName` set to `bsmd`.
+If the reload of the configuration is done an event of type `uei.opennms.org/internal/reloadDaemonConfigSuccessful` is fired.
+The following options to trigger a configuration reload are available:
+
+- Use the `Reload` button on the `Manage Business Services` page in the administration section of the web UI.
```
Comment (indigo423) [inline opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc]: Sometimes a bullet point is a sentence with "." sometimes not and most have just one bullet point.

File: opennms-doc/guide-development/src/asciidoc/text/bsm/bsm-development.adoc [hunk 15]
```diff
@@ -0,0 +1,147 @@
+
+// Allow GitHub image rendering
+:imagesdir: ../../../images
+
+==== Business Service Monitoring
```
Comment (indigo423) [inline opennms-doc/guide-development/src/asciidoc/text/bsm/bsm-development.adoc]: please add new lines after heading to be consistent

File: opennms-doc/guide-development/src/asciidoc/text/bsm/bsm-development.adoc [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+            <type>Identity</type>
+          </map-function>
+          <weight>1</weight>
+        </child>
+      </child-edges>
+      <reduction-function><type>MostCritical</type></reduction-function>
+    </business-service>
+----
+
+The whole model is defined in `jetty-webapps/opennms/WEB-INF/lib/org.opennms.features.bsm.rest.api-*.jar` which can be
```
Comment (indigo423) [inline opennms-doc/guide-development/src/asciidoc/text/bsm/bsm-development.adoc]: one line one sentence.

File: opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc (hunk 2 above)
Comment (fooker) [inline opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc]: don't you think that ONE comment is enough for the whole document?

File: opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc (hunk 4 above)
Comment (fooker) [inline opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc]: don't you think that ONE comment is enough for the whole document?

File: opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc (hunk 5 above)
Comment (fooker) [inline opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc]: don't you think that ONE comment is enough for the whole document?

File: opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc (hunk 6 above)
Comment (fooker) [inline opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc]: don't you think that ONE comment is enough for the whole document?

File: opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc (hunk 7 above)
Comment (fooker) [inline opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc]: don't you think that ONE comment is enough for the whole document?

File: opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc (hunk 8 above)
Comment (fooker) [inline opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc]: don't you think that ONE comment is enough for the whole document?

File: opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc (hunk 10 above)
Comment (fooker) [inline opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc]: don't you think that ONE comment is enough for the whole document?

File: opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc (hunk 11 above)
Comment (fooker) [inline opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc]: don't you think that ONE comment is enough for the whole document?

File: opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc (hunk 12 above)
Comment (fooker) [inline opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc]: don't you think that ONE comment is enough for the whole document?

**Comment by indigo423** [comment]
@fooker ok I just tried to be nice and tell you all the places, cause I'm reading it. I let you search the parts next time yourself :)

File: opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc [hunk 17]
```diff
@@ -0,0 +1,134 @@
+
+// Allow GitHub image rendering
+:imagesdir: ../../../images
+
+_OpenNMS_ provides a very powerful and fine grained service assurance layer based on the _Pollerd_ daemon.
```
Comment (indigo423) [inline opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc]: service assurance "layer" versus service assurance framework or mechanism?

File: opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+The service assurance layer monitors and tracks the status of nodes, interfaces and services.
```
Comment (indigo423) [inline opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc]: you should introduce the term "managed entity" here somewhere.

File: opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc [hunk 19]
```diff
@@ -0,0 +1,134 @@
...
+
+The _Business Service Monitoring_ (BSM) feature adds the ability to model relationships between managed entities.
+It combines sets of managed entities into hierarchically structured _Business Services_ that represent a high level view on its elements.
+The _Business Service'_ operational status is calculated from the alarm states of its associated elements determined by the service assurance layer.
+Of course, the operational status of a _Business Service_ can be checked at any time.
+If the operational status of a _Business Service_ is changing corresponding events are fired.
+
+==== Structure of a Business Service
+
+Business Services can contain attributes, _Reduction Keys_, references to monitored services and/or other business services.
```
Comment (indigo423) [inline opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc]: Business Services vs business services at the end

File: opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc [hunk 20]
```diff
@@ -0,0 +1,134 @@
...
+
+===== Reduction Keys as part of a Business Service
+
+_Business Services_ enables users to correlate the alarm based status of managed entities into its operational status.
+_Business Services_ can contain a list of _Reduction Keys_ that are used to identify the relevant alarms to summarize.
+_Reduction Keys_ can be added to business services as plain text.
```
Comment (indigo423) [inline opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc]: Business Services vs. business services

File: opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc [hunk 21]
```diff
@@ -0,0 +1,134 @@
...
+_IP Services_ in a business service are interpreted as two _Reduction Keys_:
+
+* `uei.opennms.org/nodes/nodeLostService::%nodeID%:%ipAddress%:%serviceName%`
+-- This _Reduction Key_ matches the alarms fired if the monitored service goes down.
+* `uei.opennms.org/nodes/nodeDown::%nodeID%`
+-- This _Reduction Key_ matches the alarms fired if the entire node providing the monitored-service goes down.
+
+====== Business Services as part of Business Services
+
+_Business services_ can form relationships woth other _Business Services_.
```
Comment (indigo423) [inline opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc]: Business services vs. Business Services vs. business services

Comment (indigo423) [inline opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc]: woth vs. with

File: opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc (hunk 13 above)
Comment (fooker) [inline opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc]: don't you think that ONE comment is enough for the whole document?

File: opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc [hunk 22]
```diff
@@ -0,0 +1,134 @@
...
+
+Every _Business Service_ maintains an operational status that represents the overall status of all the contained elements.
+The set of status are similar to the severities defined by _OpenNMS_ itself.
+
+.States
+[options="header, autowidth"]
+|===
+| Name            | Numerical code | Color / Code         | Description
+| `Critical`      | _7_            | Purple      / `#c00` | This event means that a severe service affecting event has occured.
+| `Major`         | _6_            | Red         / `#f30` | Indicates seriuos disruption or malfunction of a service or system.
```
Comment (indigo423) [inline opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc]: seriuos vs. serious

File: opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc [hunk 23]
```diff
@@ -0,0 +1,134 @@
...
+| `Minor`         | _5_            | Orange      / `#f90` | Used for troubles that have not immediate effect on service or system performance.
+| `Warning`       | _4_            | Yellow      / `#fc0` | An event has occurred that may require action.
+                                                            This severity can also be used to indicate a condition that should be noted (logged) but does not require immediate action.
+| `Normal`        | _3_            | Dark green  / `#360` | Informational message. No action required.
```
Comment (indigo423) [inline opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc]: separate line, can also be one sentence.

File: opennms-doc/guide-development/src/asciidoc/text/bsm/bsm-development.adoc [hunk 24]
```diff
@@ -0,0 +1,149 @@
+
+// Allow GitHub image rendering
+:imagesdir: ../../../images
+
+==== Business Service Monitoring
+
+Every aspect of the _Business Service Monitoring_ feature can be controlled via a ReST API.
+The API's endpoint for managing _Business Services_ is located at `/opennms/api/v2/business-services`.
+It supports _XML_ content to represent the _Business Services_. The responses generated by the ReST API
```
Comment (indigo423) [inline opennms-doc/guide-development/src/asciidoc/text/bsm/bsm-development.adoc]: one line one sentence

File: opennms-doc/guide-development/src/asciidoc/text/bsm/bsm-development.adoc [hunk 25]
```diff
@@ -0,0 +1,149 @@
...
+        <type>Identity</type>
+      </map-function>
+      <weight>1</weight>
+    </child>
+  </child-edges>
+  <reduction-function><type>MostCritical</type></reduction-function>
+</business-service>
+----
+
+The whole model is defined in `jetty-webapps/opennms/WEB-INF/lib/org.opennms.features.bsm.rest.api-*.jar` which can be
```
Comment (indigo423) [inline opennms-doc/guide-development/src/asciidoc/text/bsm/bsm-development.adoc]: one line one sentence

File: opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc (hunk 18 above)
Comment (fooker) [inline opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc]: i.e.: a managed entity is an entity that is managed by an entity manager which manages all the managed entities.

File: opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc [hunk 26]
```diff
@@ -0,0 +1,136 @@
...
+* `uei.opennms.org/nodes/nodeLostService::%nodeID%:%ipAddress%:%serviceName%`
+This _Reduction Key_ matches the alarms fired if the monitored service goes down.
+* `uei.opennms.org/nodes/interfaceDown::%nodeID%:%ipAddress%`
+This _Reduction Key_ matches the alarms fired if the interface with the given IP address on the node goes down.
+* `uei.opennms.org/nodes/nodeDown::%nodeID%`
+This _Reduction Key_ matches the alarms fired if the entire node providing the monitored service goes down.
+
+==== Business Services
+
+_Business Services_ can form relationships woth other _Business Services_.
```
Comment (mvrueden) [inline opennms-doc/guide-admin/src/asciidoc/text/bsm/bsm-admin.adoc]: Is `woth` but should be `with`

**Pull request merged**